collect_app/src/main/java/org/odk/collect/android/formentry/FormEntryViewModel.java

```diff
@@ -369,14 +370,12 @@ public void validate() {
                     }
 
                     // JavaRosa moves to the index where the contraint failed
-                    if (result instanceof FailedValidationResult) {
+                    if (result instanceof FailedValidationResult && !((FailedValidationResult) result).getIndex().equals(originalIndex)) {
                         updateIndex(true);
                     }
-
-                    return result;
-                }, result -> {
-                    validationResult.setValue(new Consumable<>(result));
-                }
+                    validationResult.postValue(new Consumable<>(result));
```

Comment:
Nice find! 

I'm not confident this change will always fix the problem if we have a potentially race condition here: `setValue` in the foreground and `postValue` in the background are semantically very similar as far as I can understand (as `postValue` will just do everything `setValue` does but on a future UI thread loop). It feels to me like we're at risk of recreating this bug in the future.

I'm wondering if instead we should rework how we propagate validation errors to the UI from the ViewModel so the UI is not reacting to multiple events and creating a risk of them overlapping badly. I'm imagining that `getCurrentIndex` could be of type `LiveData<Pair<FormIndex, FailedValidationResult?>>` so that the view that the error can simply be set as part of us building the view (this would mean removing the nice optimisation you've added around not re-rendering for the same index sadly). Alternatively, `FormEntryViewModel` could expose a `getValidationErrorForIndex(index: FormIndex)` method that returns an error for the index if there is one and could be called for each question while creating the view. We'd still need a `getValidationResult()` method for showing the success toast, but it's fine to have overlapping (and consumable) events there as the UI it's interacting with is different.

Reply:
Would you like this more: https://github.com/getodk/collect/commit/3325413ad0ab8831a32446896702151a3c5af956
I'm not sure so you can decide what's better.

Reply:
> Would you like this more: https://github.com/getodk/collect/commit/3325413ad0ab8831a32446896702151a3c5af956

Yeah I like that approach a lot more. Now we get both pieces of information (the index and the optional error) at the same time and have more control how we react to that in the UI.